src/loop_logic.rs

```diff
@@ -358,23 +358,58 @@ impl<'l, Data> EventLoop<'l, Data> {
     }
 
     fn dispatch_idles(&mut self, data: &mut Data) {
-        let idles = ::std::mem::take(&mut *self.handle.inner.idles.borrow_mut());
+        let idles = std::mem::take(&mut *self.handle.inner.idles.borrow_mut());
         for idle in idles {
             idle.borrow_mut().dispatch(data);
         }
     }
 
     fn invoke_pre_run(&self, data: &mut Data) -> crate::Result<()> {
-        for (_, source) in self.handle.inner.sources.borrow().iter() {
+        let keys = self
+            .handle
+            .inner
+            .sources
+            .borrow()
+            .keys()
+            .collect::<Vec<_>>();
+
+        for key in keys {
+            let source = self
+                .handle()
+                .inner
+                .sources
+                .borrow()
+                .get(key)
+                .unwrap()
```

Comment:
Running into a `None` unwrap panic here quite often:

```
[2022-12-05][12:37:53][panic][ERROR] thread 'main' panicked at 'called `Option::unwrap()` on a `None` value': .cargo/registry/src/github.com-1ecc6299db9ec823/calloop-0.10.3/src/loop_logic.rs:383
   0: <backtrace::capture::Backtrace as core::default::Default>::default
             at .cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.66/src/capture.rs:410:9
   1: {closure#0}
             at .cargo/registry/src/github.com-1ecc6299db9ec823/log-panics-2.1.0/src/lib.rs:115:29
   2: std::panicking::rust_panic_with_hook
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:702:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:586:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/sys_common/backtrace.rs:138:18
   5: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   6: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   7: core::panicking::panic
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:48:5
   8: core::option::Option<T>::unwrap
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/option.rs:775:21
   9: invoke_pre_run<winit::platform_impl::platform::wayland::event_loop::state::WinitState>
             at .cargo/registry/src/github.com-1ecc6299db9ec823/calloop-0.10.3/src/loop_logic.rs:377:26
  10: calloop::loop_logic::EventLoop<Data>::dispatch
             at .cargo/registry/src/github.com-1ecc6299db9ec823/calloop-0.10.3/src/loop_logic.rs:429:9
  11: winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::loop_dispatch
             at .cargo/git/checkouts/winit-c2fdb27092aba5a7/a7a7cc6/src/platform_impl/linux/wayland/event_loop/mod.rs:578:9
  12: run_return<(), breda_render_loop::{impl#0}::run_loop::{closure_env#0}<core::result::Result<(), anyhow::Error>>>
             at .cargo/git/checkouts/winit-c2fdb27092aba5a7/a7a7cc6/src/platform_impl/linux/wayland/event_loop/mod.rs:286:41
  13: winit::platform_impl::platform::EventLoop<T>::run_return
             at .cargo/git/checkouts/winit-c2fdb27092aba5a7/a7a7cc6/src/platform_impl/linux/mod.rs:754:56
  14: <winit::event_loop::EventLoop<T> as winit::platform::run_return::EventLoopExtRunReturn>::run_return
             at .cargo/git/checkouts/winit-c2fdb27092aba5a7/a7a7cc6/src/platform/run_return.rs:51:9
```

Reply:
Oh, could you try https://github.com/Smithay/calloop/pull/117 if it fixes it for you?

Reply:
Yes, that fixes it, thanks!